webapp/src/app/participant-view/participant-stage-view/exp-leader-reveal/exp-leader-reveal.component.ts

```diff
@@ -1,80 +1,66 @@
-import { Component, Inject, signal, Signal } from '@angular/core';
+import { Component, Inject, Input, signal, Signal } from '@angular/core';
+import { MatButtonModule } from '@angular/material/button';
+import { injectQueryClient } from '@tanstack/angular-query-experimental';
 
 import { ProviderService } from 'src/app/services/provider.service';
+import { updateLeaderRevealStageMutation } from 'src/lib/api/mutations';
 import { Participant } from 'src/lib/participant';
 import { PARTICIPANT_PROVIDER_TOKEN } from 'src/lib/provider-tokens';
-import { StageKind } from 'src/lib/types/stages.types';
+import { ExpStageVoteReveal } from 'src/lib/types/stages.types';
 import { VoteReveal } from 'src/lib/types/votes.types';
 
 @Component({
   selector: 'app-exp-leader-reveal',
   standalone: true,
-  imports: [],
+  imports: [MatButtonModule],
   templateUrl: './exp-leader-reveal.component.html',
   styleUrl: './exp-leader-reveal.component.scss',
 })
 export class ExpLeaderRevealComponent {
+  // Reload the internal logic dynamically when the stage changes
+  @Input({ required: true })
+  set stage(value: ExpStageVoteReveal) {
+    this._stage = value;
+
+    this.stageData = this.stage.config;
+    this.everyoneReachedThisStage = this.participant.everyoneReachedCurrentStage(this.stage.name);
+  }
+
+  private queryClient = injectQueryClient();
+  private mutationReveal = updateLeaderRevealStageMutation(this.queryClient, () =>
+    this.participant.navigateToNextStage(),
+  );
+
+  get stage(): ExpStageVoteReveal {
+    return this._stage as ExpStageVoteReveal;
+  }
+
+  private _stage?: ExpStageVoteReveal;
+
   public participant: Participant;
   public stageData: VoteReveal;
 
-  public everyoneReachedTheEnd: Signal<boolean>;
+  public everyoneReachedThisStage: Signal<boolean>;
   public finalLeader: Signal<string>;
 
   constructor(
     @Inject(PARTICIPANT_PROVIDER_TOKEN) participantProvider: ProviderService<Participant>,
   ) {
     this.participant = participantProvider.get();
-    const { config } = this.participant.assertViewingStageCast(StageKind.RevealVoted)!;
-
-    this.stageData = config;
+    this.stageData = this.stage?.config; // This will truly be initialized in ngOnInit. this.stage can be undefined here
 
-    // TODO: use the new backend way of syncing participant progression
-    this.everyoneReachedTheEnd = signal<boolean>(false);
-    //  computed(() => {
-    //   const users = Object.values(this.participant.experiment().participants);
-    //   const isReady = users
-    //     .map((userData) => userData.futureStageNames.length)
-    //     .every((n) => n === 1);
-    //   // Allow "Next" to be pushed.
-    //   if (isReady) {
-    //     for (const user of users) {
-    //       user.allowedStageProgressionMap[user.workingOnStageName] = true;
-    //     }
-    //   }
-    //   return isReady;
-    // });
+    this.everyoneReachedThisStage = signal<boolean>(false);
```

Comment:
Maybe add a comment to say that this is signal is expected to be thrown away and replaced OnInit when the stage data is set... ?

Reply:
Might be better to simply not have this initialized... otherwise you risk creating a dependency on the signal and then having the signal value replaced, but the dependency update forgotten about... Alternatively, whenever the stage is set, you write updates from the stage's signal to this signal... (that might be better, then you have the value always defined)